src/ir_visitor/visitor_data.rs

```diff
@@ -0,0 +1,62 @@
+use crate::cmdline;
+use fil_ir as ir;
+use ir::MutCtx;
+
+/// Trait for visitor data passed to visitor functions in an [ir::Visitor] implementation.
+pub trait VData<'comp> {
+    fn new(
+        idx: ir::CompIdx,
+        opts: &'comp cmdline::Opts,
+        ctx: &'comp mut ir::Context,
+    ) -> Self;
+
+    fn comp(&'comp mut self) -> &'comp mut ir::Component;
+}
+
+/// Contains information passed to visitor functions.
+/// This is re-generated when beginning to visit each component.
+pub struct VisitorData<'comp> {
+    /// The current component being visited
+    pub comp: ir::Component,
+    /// The idx of the current component.
+    pub idx: ir::CompIdx,
+    /// The command line options for this pass
+    pub opts: &'comp cmdline::Opts,
+    /// mutable context reference, held to prevent another
+    /// function from mutating the context as it is currently invalid.
+    pub mut_ctx: &'comp mut ir::Context,
+}
+
+impl<'comp> VisitorData<'comp> {
+    /// Get an immutable reference to the current [ir::Context].
+    pub fn ctx(&'comp self) -> &'comp ir::Context {
+        self.mut_ctx
+    }
+}
+
+impl<'comp> VData<'comp> for VisitorData<'comp> {
+    fn comp(&mut self) -> &mut ir::Component {
+        &mut self.comp
+    }
+
+    fn new(
+        idx: ir::CompIdx,
+        opts: &'comp cmdline::Opts,
+        ctx: &'comp mut ir::Context,
+    ) -> Self {
+        let comp = std::mem::take(ctx.get_mut(idx));
+        Self {
+            comp,
+            idx,
+            opts,
+            mut_ctx: ctx,
+        }
+    }
+}
+
+impl Drop for VisitorData<'_> {
```

Comment:
When you find yourself implementing drop, something has gone wrong :rofl:

Reply:
Funnily enough that's the only thing that's been there since before this change - I added it way back when visitordata was first introduced.